src/array/cuda/rowwise_sampling.cu

```diff
@@ -232,27 +233,372 @@ __global__ void _CSRRowWiseSampleReplaceKernel(
 
     if (deg > 0) {
       // each thread then blindly copies in rows only if deg > 0.
-      for (int idx = threadIdx.x; idx < num_picks; idx += CTA_SIZE) {
+      for (int idx = threadIdx.x; idx < num_picks; idx += BLOCK_SIZE) {
         const int64_t edge = curand(&rng) % deg;
         const int64_t out_idx = out_row_start+idx;
         out_rows[out_idx] = row;
         out_cols[out_idx] = in_index[in_row_start+edge];
         out_idxs[out_idx] = data ? data[in_row_start+edge] : in_row_start+edge;
       }
     }
-    out_row += BLOCK_CTAS;
+    out_row += 1;
+  }
+}
+
+/**
+* @brief Equivalent to numpy expression: array[idx[off:off + len]]
+*
+* @tparam IdType The ID type used for indices.
+* @tparam FloatType The float type used for array values.
+* @param array The array to be selected.
+* @param idx_data The index mapping array.
+* @param index The index of value to be selected.
+* @param offset The offset to start.
+* @param out The selected value (output).
+*/
+template<typename IdType, typename FloatType>
+__device__ void _DoubleSlice(
+    const FloatType * const array,
+    const IdType * const idx_data,
+    const IdType idx,
+    const IdType offset,
+    FloatType* const out) {
+  if (idx_data) {
+    *out = array[idx_data[offset+idx]];
+  } else {
+    *out = array[offset+idx];
+  }
+}
+
+/**
+* @brief Perform row-wise sampling on a CSR matrix, and generate a COO matrix,
+* without replacement.
+*
```

Comment:
Can we add 2-3 lines of description of the design choices (e.g., parallelization strategy, use of shared mem, etc) made in the implementation?

Reply:
Updated accordingly.